index.bs

```diff
@@ -1882,6 +1884,55 @@ To <dfn>get the navigation info</dfn>, given |context| and |navigation status|:
 
 ### Commands ### {#module-browsingContext-commands}
 
+#### The browsingContext.create Command ####  {#command-browsingContext-create}
+
+The <dfn export for=commands>browsingContext.create</dfn> command creates a new
+[=/browsing context=], either in a new window or in a new tab, and returns its
+[=browsing context id=].
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      BrowsingContextCreateCommand = {
+        method: "browsingContext.create",
+        params: BrowsingContextCreateParameters
+      }
+
+      BrowsingContextCreateType = "tab" / "window"
+
+      BrowsingContextCreateParameters = {
+        type: BrowsingContextCreateType
+      }
+      </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+        BrowsingContextCreateResult = {
+          context: BrowsingContext
+        }
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for browsingContext.create">
+The [=remote end steps=] with |command parameters| are:
+
+  1. Let |type| be the value of the <code>type</code> field of
+     |command parameters| if present, or "<code>window</code>" otherwise.
+
+  1. TODO: insert step 5 of https://w3c.github.io/webdriver/#new-window verbatim here?
```

Comment:
This is really the main question of this PR, everything I've done is mostly the boilerplate.

https://w3c.github.io/webdriver/#new-window involves the "current browsing context" which we probably want to avoid in BiDi. But without that, presumably we should pass in another browsing context which the new one should be opened after in the tab case?

Reply:
Yeah I guess it makes sense to have an optional context, although absent that there's probably a way to talk about using the browsing context which has, or most recently had, OS-level focus as the relevant context in the tab case.

It occurs to me this should be accompanied by a PR to provide a switch-to-context method c.f. https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-bringToFront

Reply:
Calling CDP's Target.createTarget is a lot like a user opening a new unrelated tab, in which case there's no need for an existing/current browsing context.

It almost feels like we should have two commands; One that invokes the `window.open` steps as WebDriver/HTTP's New Window command does, and one that just creates a new tab/window without an opener relationship as Target.createTarget does.

Reply:
Aha, so rather than a `type` argument, two commands for these two operations? https://chromedevtools.github.io/devtools-protocol/tot/Target/#method-createTarget does have a `newWindow` boolean though, so the precedent seems to be a single command, both in CDP and WebDriver. Am I misunderstanding the suggestion?

Reply:
@jgraham regarding `Page.bringToFront` in CDP, do you think the equivalent in BiDi should be `browsingContext.bringToFront` then?

Perhaps we need to make the concept of a group of browsing contexts (tabs) explicit? Playwright has a notion of [`BrowserContext`](https://playwright.dev/docs/api/class-browsercontext) which seems to group together their `Page` objects. The main reason for it seems to be "independent browser sessions" rather than representing a visible window, but there might be something to this?

Reply:
Note that WebDriver is creating the window with `noopener`, so it's really a new tab that's not connected to any other tabs. So I think one command that opens a new top level browsing context in either a tab or a window should be fine.

I think BrowserContext in Playwrihgt is trying to solve a different kind of problem; I think it's an abstraction over the idea of multiple containers (in gecko) or private windows (in Blink), which don't share state. It seems like a reasonable idea to surface that concept, but I don't think it's apping 1:1 with OS windows, and is probably a more advanced feature. It's also not a 1:1 mapping with OS windows, which I think is what you're looking for here.

For my money, a `browsingContext.bringToFront` that would take the context parameter, get the associated top-level-browsing context, and do [something] makes sense. The question is whether [something] should require that it ends up with OS-level focus or just that it's selected context in whatever window it's in without requiring anything at the OS level.

Reply:
OK, let's put Playwright's `BrowserContext` to the side for now.

For the purposes of this PR, my main question is if we want a `browsingContext.create` command to work without any reference to which window it should open the new browsing context in. The slightly unsatisfying bit is going to be if we add a notion of windows later, there would have to be an implicit window here and we've tried to avoid things like WebDriver's "current top-level browsing context" in BiDi so far.

Reply:
I think my proposal from https://github.com/w3c/webdriver-bidi/pull/133#discussion_r704346287 still makes sense: if you provide a context, the new context should be created as if the user was creating a new context starting at that one (that's hard to specify in detail because to some extent it's an implementation detail, but roughly it would mean a new tab would share an os window with the tab containing that context; I think the text will be necessarily handwavy here).

Reply:
I've left this as an inline issue in the spec now, reading:

>  HTML’s window open steps depend on the "entry global object" and WebDriver’s New Window command depends on the "current browsing context", but we have neither here. What OS window should new tabs be opened in?